### DISCUSSION TITLE
How does sqlalchemy implement the expression evaluation?

### DISCUSSION THREAD

**Discussion by xiaojiuwo1993** [sqlalchemy/sqlalchemy#11817] [Usage Questions] [answered]
when we fetch data, we use .filter(class.property1 == value1, class.property2 == value2)
The defination of filter in src of sqlarchmy is 
```def filter(self, *criterion: _ColumnExpressionArgument[bool]) -> Self:```
However when I write my code to implement what the filter does, it just not work good
```class User:
    name: str = None
    age: int = None


def filter(*criterion: list):
    criterion
    pass


filter(User.name == "python", User.age == 10)
```
Ideally the criterion should be get as a expression like ``` ['User.name ==  "python"', 'User.age == 10']``` but it only give the result ```[0,0]```
In C# there are lambda expression, you can evaluate the expression tree in runtime. Would the community simply explain about how this work?

**Comment by CaselIT** [comment]
Hi,

I don't understand the question. Are you trying to re-implement your own evaluator? If so you will need to override __eq__ of your type to return something that's not a bool

**Reply by xiaojiuwo1993** [reply]
Yes I want to re-implement evaluator like sqlarchemy, but do not inherit from sqlarchemy classes. Would you mind give me a hint about the implementation?

**Reply by CaselIT** [reply, marked as answer]
The general idea is illustrated here https://realpython.com/operator-function-overloading/#overloading-built-in-operators

In sqlalchemy the overload are defined here https://github.com/sqlalchemy/sqlalchemy/blob/706b1f77b84cc921dbab54a38f73bc40378ecb86/lib/sqlalchemy/sql/operators.py#L496

**Reply by xiaojiuwo1993** [reply]
What an amazing use of the magic method!!!!!!!!!!!!!!!!!!
